### DISCUSSION TITLE
How to permanently update MODELICAPATH or OEPNMODELICALIBRARY via CLI?

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#11883] [Q&A]
When I start the OMEdit the `MODELICAPATH ` displayed in `Tools -> Options -> Library` is identical to the path `OPENMODLICALIBRARY` on windows env variables.

Then i use CLI to update the path with `setModelicaPath(getModelicaPath()+"new_path")` and it seems to work just fine since i can now find the library in `new_path` and choose it as system library. The prbolem is, when i closed and restarted OMEdit, I got an error stating `failed to load package using <list of paths>` and some how `new_path` was not in `<list of paths>` 

so how does one should permanently update their  `MODELICAPATH` or `OEPNMODELICALIBRARY`?

**Comment by casella** [comment]
If you use OMEdit, you can change the setting in Tools | Options | Library. That would be reapplied next time you open it. The changes made to the CLI are only applied to the currently opened session.

**Reply by Arinomo** [reply]
@casella thank you for you response.
I changed the `MODELICAPATH` from Tools | Options | Library and got following error.
```
[1] 08:29:59 Translation Notification
[C:/OpenModelica/share/omlibrary/cache/index.json: 0:0-0:0]: Cached libraries were found and will be installed into //<userpath>/Anwendungsdaten/.openmodelica/libraries/.

[2] 08:29:59 Scripting Error
[//<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package Complex that provides version 4.0.0+maint.om. The following versions are available:

[3] 08:29:59 Scripting Error
[//<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package Modelica that provides version 3.2.3+maint.om. The following versions are available:

[4] 08:29:59 Scripting Error
[//<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package Modelica that provides version 4.0.0+maint.om. The following versions are available:

[5] 08:29:59 Scripting Error
[//<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package ModelicaReference that provides version 4.0.0+maint.om. The following versions are available:

[6] 08:29:59 Scripting Error
[//<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package ModelicaServices that provides version 4.0.0+maint.om. The following versions are available:

[7] 08:29:59 Scripting Error
The MODELICAPATH (<newpath>/Modelica 4.0.0+maint.om) does not contain //<userpath>/Anwendungsdaten/.openmodelica/libraries/, so the package index cannot be used.

[8] 08:29:59 Scripting Error
[//<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package ObsoleteModelica4 that provides version 4.0.0+maint.om. The following versions are available:

[9] 08:29:59 Error
Curl error for URL https://libraries.openmodelica.org/index/v1/index.json: Timeout was reached

[10] 08:29:59 Scripting Error
Failed to download package index https://libraries.openmodelica.org/index/v1/index.json to file //<userpath>/Anwendungsdaten/.openmodelica/libraries/index.json.

```

`<newpath>` is where the `Modelica 4.0.0+maint.om`, `ModelicaReference 4.0.0+maint.om` and `ModelicaServices 4.0.0+maint.om` located. The reason we put it in `<newpath>` and not `~/.openmodelica/libraries` is because we running the OMEditor via CITRIX and as far as i know, it wont have any internet connection, so OMEdit could not access the online repos.

**Reply by casella** [reply]
The Windows installer automatically installs MSL and related libraries from a cache if there is no internet connection. I don't remember how it works with Linux, maybe @sjoelund can comment on that.

**Reply by olivleno** [reply]
How to add a list of paths in the OMEdit options dialog under Windows?

**Reply by casella** [reply]
@adeas31 can you please comment on that?

**Comment by adrpo** [comment]
I don't think is yet possible from the GUi, but it is really easy to add. I implemented that for the CRML dialog where you can select multiple paths.

You can do it from the CLI via:

`setModelicaPath(getModelicaPath() + ";your/path");`

Note that you need `;` on Windows and `:` on Linux as path separators (I think we have a script function to get the proper separator, I will check).

**Reply by adeas31** [reply]
It is possible from the GUI. Set `MODELICAPATH` in `Tools | Options | Library`. Multiple paths are separated by `;` on Windows and `:` on Linux.

**Reply by olivleno** [reply]
Thanks for the hint, which I should have found it myself looking at the tool tip. Sorry about that.

It seems that all variants:
`C:\Users\xxx\AppData\Roaming\.openmodelica\libraries; C:\PROGRA~2\TLK\Modelica`
`C:/Users/xxxx/AppData/Roaming/.openmodelica/libraries; C:/Program Files (x86)/TLK/Modelica`
are accepted without error message. That's good.

Never-the-less I fail to load a model using the TIL library, despite of having extended the Modelica path.
```
[1] 11:07:48 Scripting Notification
Automatically loaded package Complex 4.0.0 due to uses annotation from Modelica.

[2] 11:07:48 Scripting Notification
Automatically loaded package ModelicaServices 4.0.0 due to uses annotation from Modelica.

[3] 11:08:27 Scripting Notification
Skipped loading package TIL (2025.2) using MODELICAPATH C:/Users/leo2rng/AppData/Roaming/.openmodelica/libraries; C:/Program Files (x86)/TLK/Modelica (uses-annotation may be wrong).

[4] 11:08:27 Scripting Notification
You can install the requested package using one of the commands:
  installPackage(TILMedia, "main", exactMatch=true)
  installPackage(TILMedia, "2025.2", exactMatch=false).

[5] 11:08:27 Scripting Notification
Skipped loading package TILMedia (2025.2) using MODELICAPATH C:/Users/leo2rng/AppData/Roaming/.openmodelica/libraries; C:/Program Files (x86)/TLK/Modelica (uses-annotation may be wrong).
```

**Comment by adrpo** [comment]
There is a space in your MODELICAPATH somehow right after ";"

**Reply by adeas31** [reply]
Me and @perost tested this a bit. Space in `MODELICAPATH` causes the issue.
Also we noticed that it fails if you have space in your version string e.g.,

```modelica
model M
  TILMedia.VLEFluid vleFluid annotation(
    Placement(transformation(origin = {-14, 24}, extent = {{-10, -10}, {10, 10}})));
annotation(uses(TILMedia(version = "1.8.1 ClaRa")));
end M;
```

This works `annotation(uses(TILMedia(version = "1.8.1-ClaRa")));` or even `annotation(uses(TILMedia(version = "1.8.1")));`

**Reply by olivleno** [reply]
Thanks for investigating.
The model contains:
```
  annotation (uses(
      TIL(version="2025.2"),
      TILMedia(version="2025.2"),
      Modelica(version="4.0.0")));
```
The MODELICAPATH is now avoiding blanks:`C:\Users\xxx\AppData\Roaming\.openmodelica\libraries; C:\PROGRA~2\TLK\Modelica`.
The libraries are stored in subfolders like `TILMedia 2025.2` are now renamed to `TILMedia_2025.2`.

After loading the model I'll get:
```
[1] 08:59:05 Scripting Notification
Automatically loaded package Complex 4.0.0 due to uses annotation from Modelica.

[2] 08:59:05 Scripting Notification
Automatically loaded package ModelicaServices 4.0.0 due to uses annotation from Modelica.

[3] 08:59:12 Scripting Notification
Skipped loading package TIL (2025.2) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries; C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).

[4] 08:59:12 Scripting Notification
You can install the requested package using one of the commands:
  installPackage(TILMedia, "main", exactMatch=true)
  installPackage(TILMedia, "2025.2", exactMatch=false).

[5] 08:59:12 Scripting Notification
Skipped loading package TILMedia (2025.2) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries; C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).
```

What does the Modelica specification say regarding separator between library name and version in uses annotation?
What separator does OMC expect regarding the library folder name: `<lib_name>"-"<version>` ?

**Reply by adeas31** [reply]
> The MODELICAPATH is now avoiding blanks:`C:\Users\xxx\AppData\Roaming\.openmodelica\libraries; C:\PROGRA~2\TLK\Modelica`.

We meant the space after the separator `;`. It should be `C:\Users\xxx\AppData\Roaming\.openmodelica\libraries;C:\PROGRA~2\TLK\Modelica` instead of `C:\Users\xxx\AppData\Roaming\.openmodelica\libraries; C:\PROGRA~2\TLK\Modelica`

The space in the path works fine e.g., `C:/Program Files (x86)/TLK/Modelica`

**Reply by olivleno** [reply]
Ok, good to know. Now I reverted the subfolder names as before now again containing the blank character between library name and version.
After removing the blankg after ";" I'm able to successfully load the TIL library, but loading the TILMedia still fails based on this uses annotation in TIL:
```
   uses(Modelica(version="4.0.0"),
      TILMedia(version="2025.2"),
      TIL3_AddOn_HD(version="2025.2 HD")),
```
This is the error message:
```
[3] 11:42:48 Scripting Notification
Automatically loaded package TIL 2025.2 HD due to uses annotation from xyz.

[4] 11:42:48 Scripting Notification
Skipped loading package TIL3_AddOn_HD (2025.2 HD) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries;C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).

[5] 11:42:48 Scripting Notification
You can install the requested package using one of the commands:
  installPackage(TILMedia, "main", exactMatch=true)
  installPackage(TILMedia, "2025.2", exactMatch=false).

[6] 11:42:48 Scripting Notification
Skipped loading package TILMedia (2025.2) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries;C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).
```

How comes that uses `TIL(version="2025.2")` works but `TILMedia(version="2025.2")` fails?

Could it be due to licensing of the TIL Suite?

**Reply by adeas31** [reply]
Add `-` in the version string and it will work.

```
   uses(Modelica(version="4.0.0"),
      TILMedia(version="2025.2"),
      TIL3_AddOn_HD(version="2025.2-HD")),
```

I don't know the details of this. @perost is the best person to answer this.

**Reply by olivleno** [reply]
Ok, I'll try.
Unfortunately this cannot be done in OMEdit:
`Warning: You are changing a system library class. System libraries are always read-only. Your changes will not be saved.`
I was able to apply this change in a text editor.
But unfortunately this doesn't work either:
```
[4] 12:29:30 Scripting Notification
Skipped loading package TIL3_AddOn_HD (2025.2-HD) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries;C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).

[5] 12:29:30 Scripting Notification
You can install the requested package using one of the commands:
  installPackage(TILMedia, "main", exactMatch=true)
  installPackage(TILMedia, "2025.2", exactMatch=false).

[6] 12:29:30 Scripting Notification
Skipped loading package TILMedia (2025.2) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries;C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).
```

**Reply by olivleno** [reply]
@perost do you have a solution to this?
Does OpenModelica truely not accept blank characters in version strings?
If so, does the library folder have to be renamed accordingly also?
Is OpenModelica able to load licensed libraries, if a license manager is available?

**Reply by perost** [reply]
I created two simple packages to try and recreate the issue:
```Modelica
// TILMedia 2025.2 HD/package.mo
package TILMedia
  annotation(version = "2025.2 HD", uses(TIL3_AddOn_HD(version = "2025.2 HD")));
end TILMedia;

// TIL3_AddOn_HD 2025.2 HD/package.mo
package TIL3_AddOn_HD
  annotation(version = "2025.2 HD");
end TIL3_AddOn_HD;
```
But loading TILMedia works fine for me regardless of if the uses-annotation for `TIL3_AddOn_HD` is `2025.2` or `2025.2 HD`, and it also works both if I set `OPENMODELICALIBRARY` to only the folder containing these packages or also include my normal library folder which contains other versions of TILMedia. Is this representative of the structure you're using, or is it different in some way?

Regarding adding `-` that @adeas31 recommended, I think there was some confusion because when discussing it with me he was using `TILMedia 1.8.1-ClaRa` as an example. In your case there doesn't seem to be a `-` in the version, so adding one to the uses-annotation will of course not work.

**Reply by olivleno** [reply]
Thanks for checking. This means its not the uses annotation of the folder path name.

I figured that the libraries TIL and TILMedia are stored as `package.moe`.
Could this be the problem? Do I need a special version of TIL encrypted in a way that OpenModelica can read it?
Or is there some other installation step for the TIL libraries needed?
In Dymola everything works out of the box.

**Reply by adeas31** [reply]
> I figured that the libraries TIL and TILMedia are stored as `package.moe`.

`package.moe` is a Dymola encrypted library that is not supported by OpenModelica. We support [SEMLA](https://github.com/modelica/Encryption-and-Licensing).

> Could this be the problem? Do I need a special version of TIL encrypted in a way that OpenModelica can read it?
Or is there some other installation step for the TIL libraries needed?
In Dymola everything works out of the box.

Yes, you need TIL encrypted with SEMLA and then need special OpenModelica version with encryption support. Send an email to openmodelica@ida.liu.se if you need that one.

**Comment by olivleno** [comment]
After all I guess the simple reason is not the MODELICAPATH, not the uses annotation, but simply the fact that I should be using OpenModelica for encrypted libraries.

My suggestion for OMEdit:
1. The error message: 
```
[4] 14:31:15 Scripting Notification
Skipped loading package TIL3_AddOn_HD (2025.2 HD) using MODELICAPATH C:/Users/xxx/AppData/Roaming/.openmodelica/libraries;C:/PROGRA~2/TLK/Modelica (uses-annotation may be wrong).
```
is in this case misleading. If the loading of the library fails due to encryption, then this should be clearly stated.

2. The fact that the separator must be ";" while "; " fails should be improved such that white characters before and after the separator are skipped.
3. If the MODELICAPATH is a list, then a table would be much more user friendly than a text box, which can cause a lot of headaches as this case illustrates.

Thanks for your support to figure this out.

**Reply by adeas31** [reply]
> 2. The fact that the separator must be ";" while "; " fails should be improved such that white characters before and after the separator are skipped.
> 3. If the MODELICAPATH is a list, then a table would be much more user friendly than a text box, which can cause a lot of headaches as this case illustrates.

I will fix that in OMEdit.
